### ISSUE TITLE
POST timeouts only using AHC client backend

### ISSUE THREAD

**Issue by ktonga** [http4s/http4s#1264]
It is kind of an special case, the request is a POST without body, `Content-Length:0` must be specified, it also has not response body, useful data is provided in the `Location` response header.

http4s version is `0.15.7a`

Here is the request performed with HTTPie:

```
http --verbose  POST 'https://va.convep.liveperson.net/api/account/1234/chat/request.json?v=1&appKey=abcd' Content-Length:0
```

Which results in following response

```
POST /api/account/1234/chat/request.json?v=1&appKey=abcd HTTP/1.1
Accept: */*
Accept-Encoding: gzip, deflate
Connection: keep-alive
Content-Length: 0
Host: va.convep.liveperson.net
User-Agent: HTTPie/0.9.9



HTTP/1.1 201 Created
Content-Length: 0
Content-Type: application/json
Date: Tue, 18 Jul 2017 01:15:22 GMT
Location: https://va.convep.liveperson.net/api/account/1234/chat/H516121171561454067-1e6b301c702c4099ad5f98e6e6ef8d92K8391827
P3P: CP="NON BUS INT NAV COM ADM CON CUR IVA IVD OTP PSA PSD TEL SAM"
Server: WS
```

Here is a code snippet to reproduce the error using http4s

```scala
object Blah extends scala.App {

  import org.http4s.Uri.uri

  val http = AsyncHttpClient()

  val uri_ = uri("https://va.convep.liveperson.net/api/account/1234/chat/request.json")
  val req = Request(POST, uri_ +? ("v", "1") +? ("appKey", "abcd"))
    .putHeaders(`Content-Length`(0))
  println(req)
  val locUri = http.fetch[Uri](req) {
    case Successful(resp) =>
      println(resp)
      resp.headers
        .collectFirst { case Location(loc) => loc.parsed.uri }
        .fold[Task[Uri]](Task.fail(Error.messageError("No URI provided for bridge chat")))(Task.now)
    case failedResponse =>
      println(failedResponse)
      Task.fail(Error.messageError(s"Unexpected status creating bridge chat: ${failedResponse.status}"))
  }

  println(locUri.unsafePerformSync)
  http.shutdownNow()
}
```

What results in a timeout with the following exception:

```
Request(method=POST, uri=https://va.convep.liveperson.net/api/account/1234/chat/request.json?v=1&appKey=abcd, headers=Headers(Content-Length: 0)
[error] (run-main-0) java.util.concurrent.TimeoutException: Request timeout to va.convep.liveperson.net/162.252.73.12:443 after 30000 ms
java.util.concurrent.TimeoutException: Request timeout to va.convep.liveperson.net/162.252.73.12:443 after 30000 ms
    at org.asynchttpclient.netty.timeout.TimeoutTimerTask.expire(TimeoutTimerTask.java:43)
    at org.asynchttpclient.netty.timeout.RequestTimeoutTimerTask.run(RequestTimeoutTimerTask.java:48)
    at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:625)
    at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:700)
    at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:428)
    at java.lang.Thread.run(Thread.java:748)
```

Relevant discussion in gitter channel starts [here](https://gitter.im/http4s/http4s?at=596eb476bf7e6af22cf0698d)

**Comment by rossabaker** [comment]
I'm pretty sure the problem is related to the `InputStreamBodyGenerator(toInputStream(req.body))`.  When I replace this with a `ByteArrayBodyGenerator` with a canned body, a proper response is returned.  `toInputStream` is of course blocking I/O.

I can't reproduce this on any other server yet.  I can reproduce this on endpoints on this server that return an error, so I don't think it has anything to do with the response body.  In fact, the deadlock occurs before `onStream` is called by the async http client.

**Comment by rossabaker** [comment]
Ideally, this would use a [ReactiveStreamsBodyGenerator](https://github.com/AsyncHttpClient/async-http-client/blob/master/client/src/main/java/org/asynchttpclient/request/body/generator/ReactiveStreamsBodyGenerator.java), but we'd need a publisher for a `Process`.  We get this on master from fs2-reactive, which is why it works on master.

**Comment by rossabaker** [comment]
Fix is released in v0.15.16.
